rfc-174-consolidate-frontend-rendering-apps.md

```diff
@@ -0,0 +1,330 @@
+- Rake tasks
+
+Use `lib` rather than `app/lib` for custom code that enhances the app but doesn’t fit into a directory in `app`
+
+Under `/config`
+
+`data`
+- YAML, JSON, CSV config files. Data files should be placed here rather than under `/lib/data`.
+
+`locales`
+- Static content like lists of contact information should ideally be stored in the locale files to promote localisation.
+
+Namespaces will be used to group functionality e.g. for StepBySteps, and resources in the routes file. This structure will be documented in README of the host app.
+Each application will be analysed before being consolidated to ensure redundant code is not being moved over.
+
+### How the proposal will be carried out
+This is obviously a very large piece of work that will take some time to complete. We have identified the following steps that will be needed.
+
+* Prepare `frontend` to be the host
+    * convert frontend tests to RSpec
+    * modernise the tests
+    * audit `frontend` and make changes to the frontend directory structure to match “new layout”
+    * remove redundant code from frontend (non-blocking)
+    * remove publishing_api tasks from frontend (non-blocking)
+* Merge `feedback` into `frontend`
+    * use as a test case and use that to inform our next steps
+* Write a roadmap for app consolidation
+    * will be handled separately
+    * might be simply the order in which applications will be merged into the host
+    * or could be some apps merged together before merging into the host
+    * or some routes merging ahead of others
+* Consider and work on expected issues (see below)
+* Work to merge remaining apps
+
+### Steps as part of a merge
+We anticipate that any merge will likely involve some or all of the following steps.
+
+* Identify any issues with existing code
+    * size these issues
+    * either fix issues or record them as tech debt - we need to avoid trying to fix everything and concentrate on progressing consolidation
+* Replicate the rendering from the candidate to the host
+    * preserve the git history of the candidate app - add a permalink to the files being consolidated in the commit in the host app.
+* Change content api
+    * change the rendering app for the document type in the publishing app.
+    * republish content
+    * republish any special routes (things that don’t have a publishing app, like the homepage) - any special routes should be moved into special route publisher (we have an outstanding tech debt card relating to this)
+* Optimise the frontend code
+    * remove duplication
+* Remove the routes from the candidate
+* [Retire the candidate] once all routes have been removed
+
+### Expected issues
+Merging all of our apps together is not without difficulties and we acknowledge that a single app will present problems that our multiple apps do not.
+
+* A Rails update for one large app might be more complex than spread across individual apps.
+* The scaling of a single app might become more costly.
```

Comment:
The cost of scaling these pods is likely be minimal vs our overall infrastructure spend.